flink-end-to-end-tests/flink-streaming-kinesis-test/src/main/java/org/apache/flink/streaming/kinesis/test/KinesisExample.java

```diff
@@ -0,0 +1,156 @@
+	 */
+	interface PubsubClient {
+		void createTopic(String topic, int partitions, Properties props) throws Exception;
+
+		void sendMessage(String topic, String msg);
+
+		List<String> readAllMessages(String streamName) throws Exception;
+	}
+
+	public static void main(String[] args) throws Exception {
+		LOG.info("System properties: {}", System.getProperties());
+		// parse input arguments
+		final ParameterTool parameterTool = ParameterTool.fromArgs(args);
+		StreamExecutionEnvironment env = KafkaExampleUtil.prepareExecutionEnv(parameterTool);
+
+		String inputStream = parameterTool.getRequired("input-stream");
+		String outputStream = parameterTool.getRequired("output-stream");
+
+		PubsubClient pubsub = new KinesisPubsubClient(parameterTool.getProperties());
+		pubsub.createTopic(inputStream, 2, parameterTool.getProperties());
+		pubsub.createTopic(outputStream, 2, parameterTool.getProperties());
+
+		FlinkKinesisConsumer<KafkaEvent> consumer = new FlinkKinesisConsumer<>(
+			inputStream,
+			new KafkaEventSchema(),
+			parameterTool.getProperties());
+		consumer.setPeriodicWatermarkAssigner(new CustomWatermarkExtractor());
+
+		DataStream<KafkaEvent> input = env
+				.addSource(consumer)
+				.keyBy("word")
+				.map(new RollingAdditionMapper());
+
+		Properties producerProperties = new Properties(parameterTool.getProperties());
+		// needs region event when URL is specified
+		producerProperties.put(ConsumerConfigConstants.AWS_REGION, "us-east-1");
+
+		// KPL does not recognize endpoint URL..
+		String kinesisUrl = producerProperties.getProperty(ConsumerConfigConstants.AWS_ENDPOINT);
+		if (kinesisUrl != null) {
+			URL url = new URL(kinesisUrl);
+			producerProperties.put("KinesisEndpoint", url.getHost());
+			producerProperties.put("KinesisPort", Integer.toString(url.getPort()));
+			producerProperties.put("VerifyCertificate", "false");
+		}
+
+		FlinkKinesisProducer<KafkaEvent> producer = new FlinkKinesisProducer<>(
+			new KafkaEventSchema(),
+			producerProperties);
+		producer.setDefaultStream(outputStream);
+		producer.setDefaultPartition("fakePartition");
+
+		input.addSink(producer);
+
+		final AtomicReference<Exception> caughtException = new AtomicReference<>();
+
+		// job and following test logic need to run in parallel
+		Thread executeThread =
+			new Thread(
+				() -> {
+					try {
+						env.execute();
+						LOG.info("executed program");
+					} catch (Exception e) {
+						caughtException.set(e);
+					}
+				});
+		executeThread.start();
+
+		// generate input
+		String[] messages = { "elephant,5,45218", "squirrel,12,46213", "bee,3,51348", "squirrel,22,52444", "bee,10,53412", "elephant,9,54867" };
+		for (String msg : messages) {
+			pubsub.sendMessage(inputStream, msg);
+		}
+		LOG.info("generated records");
+
+		long timeoutMillis = System.currentTimeMillis() + 30_000;
+		List<String> results = pubsub.readAllMessages(outputStream);
+		while (System.currentTimeMillis() < timeoutMillis) {
```

Comment:
I would suggest to better use a monotonic time provider like `System.nanoTime()`. Otherwise this can lead to very rare cases of timeouts due to clock adjustment that make the test look flaky and are super hard to track/reproduce.

Reply:
Thanks for the suggestion, code also looks much cleaner using `Deadline`.